metrics/sinks/stackdriver/stackdriver.go

```diff
@@ -213,15 +217,92 @@ func (sink *StackdriverSink) ExportData(dataBatch *core.DataBatch) {
 				req.TimeSeries = append(req.TimeSeries, point)
 			}
 			if len(req.TimeSeries) >= maxTimeseriesPerRequest {
-				sink.requestQueue <- req
+				requests = append(requests, req)
 				req = getReq()
 			}
 		}
 	}
 
 	if len(req.TimeSeries) > 0 {
-		sink.requestQueue <- req
+		requests = append(requests, req)
 	}
+
+	go sink.sendRequests(requests)
+}
+
+func (sink *StackdriverSink) sendRequests(requests []*sd_api.CreateTimeSeriesRequest) {
+	// Each worker can handle at least batchExportTimeout/sdRequestLatencySec requests within the specified period.
+	// 5 extra workers just in case.
+	workers := 5 + len(requests)/(sink.batchExportTimeoutSec/sdRequestLatencySec)
+	requestQueue := make(chan *sd_api.CreateTimeSeriesRequest)
+	// Close the channel in order to cancel exporting routines.
+	defer close(requestQueue)
+
+	// Launch Go routines responsible for sending requests
+	for i := 0; i < workers; i++ {
+		go sink.requestSender(requestQueue)
+	}
+
+	timeoutTime := time.Now().Add(time.Duration(sink.batchExportTimeoutSec) * time.Second)
+
+forloop:
+	for i := range requests {
+		now := time.Now()
+		select {
+		case requestQueue <- requests[i]:
+			// yet another request added to queue
+		case <-time.After(timeoutTime.Sub(now)):
+			glog.Warningf("Timeout while exporting metrics to Stackdriver. Dropping %d out of %d requests.", len(requests)-i, len(requests))
+			// TODO(piosz): consider cancelling requests in flight
+			// Report dropped requests in metrics.
+			for j := i; j < len(requests); j++ {
+				requestsSent.WithLabelValues(strconv.Itoa(http.StatusTooManyRequests)).Inc()
+				timeseriesSent.
+					WithLabelValues(strconv.Itoa(http.StatusTooManyRequests)).
+					Add(float64(len(requests[j].TimeSeries)))
+			}
+			break forloop
+		}
+	}
+
+}
+
+func (sink *StackdriverSink) requestSender(queue chan *sd_api.CreateTimeSeriesRequest) {
+	time.Sleep(time.Duration(rand.Intn(1000*sdRequestLatencySec)) * time.Millisecond)
+	for {
+		select {
+		case req, active := <-queue:
+			if !active {
+				return
+			}
+			sink.sendOneRequest(req)
+		}
+	}
+}
+
+func (sink *StackdriverSink) sendOneRequest(req *sd_api.CreateTimeSeriesRequest) {
+	startTime := time.Now()
+	empty, err := sink.stackdriverClient.Projects.TimeSeries.Create(fullProjectName(sink.project), req).Do()
+
+	var responseCode int
+	if err != nil {
+		glog.Errorf("Error while sending request to Stackdriver %v", err)
```

Comment:
Definitely NOT an error, not even an warning, since exposed as a metric. I'd say V(2)